### ISSUE TITLE
can't recognize my programming language

### ISSUE THREAD

**Issue by ZainChen** [wakatime/vscode-wakatime#88]
Hello, wakatime can't recognize my programming language, I don't know if anyone else has encountered this problem.
![image](https://user-images.githubusercontent.com/15615524/54491965-8530bb80-48fd-11e9-8ae1-0ae31e23d667.png)

**Comment by ZainChen** [comment]
Using vscode

**Comment by alanhamlett** [comment]
Can you turn on debug mode (Press `CTRL+SHIFT+p`, type `wakatime.debug`, press `Enter`, select `true`, press `Enter`) then edit one of the files with unrecognized language. That will trigger an error message in your `C:\users\zain\.wakatime.log` file if it's a problem with the Pygments lexer parsing the file.

Otherwise, it might point us to the source of the problem.

**Comment by ZainChen** [comment]
Hello, this is debugging information for editing one of the files in the language.
But this is what I edited after setting "Custom rules" in wakeatime. I don't know if it will affect.
![image](https://user-images.githubusercontent.com/15615524/54493561-7bfb1b00-490c-11e9-9fc1-5364b12a89ba.png)
[.wakatime.log](https://github.com/wakatime/vscode-wakatime/files/2975340/default.wakatime.log)

**Comment by ZainChen** [comment]
这是"TypeScript"的debug，*.ts文件我没有"Custom rules".
[.wakatime-TypeScript.log](https://github.com/wakatime/vscode-wakatime/files/2975381/default.wakatime.log)

**Comment by alanhamlett** [comment]
From the debug logs, 3 heartbeats are sent all recognized as `TypeScript`:

```
{'time': 1552838227.101483, 'entity': 'E:/zain_git/omi-vscode/node_modules/vscode/vscode.d.ts', 'language': 'TypeScript'}
{'time': 1552838229.8758411, 'entity': 'E:/zain_git/omi-vscode/node_modules/vscode/vscode.d.ts', 'language': 'TypeScript'}
{'time': 1552838230.9471388, 'entity': 'E:/zain_git/omi-vscode/node_modules/vscode/vscode.d.ts', 'language': 'TypeScript'}
```

Custom rules don't affect client-side plugin language detection.

Maybe another file will trigger the bug? This file is detected correctly.

**Comment by ZainChen** [comment]
[.wakatime-json.log](https://github.com/wakatime/vscode-wakatime/files/2975952/default.wakatime-json.log)

**Comment by alanhamlett** [comment]
Yes, this is very helpful! The traceback shows a file encoding error:

```
Traceback (most recent call last):
  File "c:\Users\Administrator\.vscode-insiders\extensions\wakatime.vscode-wakatime-1.2.13\wakatime-master\wakatime\stats.py", line 121, in guess_lexer_using_filename
    lexer = custom_pygments_guess_lexer_for_filename(file_name, text)
  File "c:\Users\Administrator\.vscode-insiders\extensions\wakatime.vscode-wakatime-1.2.13\wakatime-master\wakatime\stats.py", line 307, in custom_pygments_guess_lexer_for_filename
    for lexer in _iter_lexerclasses():
  File "c:\Users\Administrator\.vscode-insiders\extensions\wakatime.vscode-wakatime-1.2.13\wakatime-master\wakatime\packages\pygments\lexers\__init__.py", line 233, in _iter_lexerclasses
    for lexer in find_plugin_lexers():
  File "c:\Users\Administrator\.vscode-insiders\extensions\wakatime.vscode-wakatime-1.2.13\wakatime-master\wakatime\packages\pygments\plugin.py", line 52, in find_plugin_lexers
    for entrypoint in iter_entry_points(LEXER_ENTRY_POINT):
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 645, in iter_entry_points
    entries = dist.get_entry_map(group)
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 2701, in get_entry_map
    self._get_metadata('entry_points.txt'), self
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 2408, in parse_map
    for group, lines in data:
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 3036, in split_sections
    for line in yield_lines(s):
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 2268, in yield_lines
    for ss in strs:
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 2627, in _get_metadata
    for line in self.get_metadata_lines(name):
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 1408, in get_metadata_lines
    return yield_lines(self.get_metadata(name))
  File "D:\Software\Install\NoGreenInstall\Python37\lib\site-packages\pkg_resources\__init__.py", line 1405, in get_metadata
    return value.decode('utf-8') if six.PY3 else value
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x91 in position 40: invalid start byte
```

I'll release a new version of [wakatime-cli ](https://github.com/wakatime/wakatime) soon to fix this.

**Comment by alanhamlett** [comment]
Also, what filesystem encoding is your machine using?

**Comment by ZainChen** [comment]
Are you referring to the system encoding? I use window7, enter "chcp" in the cmd.exe window to display "936 (GBK, Simplified Chinese zh_CN)", the file encoding opened by vscode is utf-8. Don't know which one is what you want, or is it the encoding look at in other ways?

**Comment by alanhamlett** [comment]
This might be a bug in setuptools library. I've opened pypa/setuptools#1722 to track it there.

**Comment by ZainChen** [comment]
Come on! I especially like wakatime.^_^

**Comment by alanhamlett** [comment]
@ZainChen can you follow the instructions from [this comment](https://github.com/pypa/setuptools/issues/1722#issuecomment-480377606) to see if a package on your system has the wrong metadata encoding?

**Comment by ZainChen** [comment]
The result of the operation is like this：
![image](https://user-images.githubusercontent.com/15615524/55711168-be50dd00-5a1e-11e9-99f1-daeefeadb752.png)

**Comment by alanhamlett** [comment]
@ZainChen thanks, but can you run the same thing again with Python3? To do that, instead of running `python` use `python3`. If `python3` isn't found, you might have to run it with the absolute path to the installed location, most likely `C:\Python37\python`.

**Comment by ZainChen** [comment]
Hello, run with python3, see if my operation is the same as you think:
![image](https://user-images.githubusercontent.com/15615524/55932935-2d684480-5c5e-11e9-893c-c243d642eb3b.png)

**Comment by gandarez** [comment]
```C:\Users\ganda\.vscode\extensions\wakatime.vscode-wakatime-1.2.14\wakatime-master>python
Python 2.7.16 (v2.7.16:413a49145e, Mar  4 2019, 01:37:19) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import wakatime
>>> import pkg_resources
>>> list(pkg_resources.iter_entry_points('pygments.lexers'))
[]

**Comment by ZainChen** [comment]
@gandarez 
Your system can't recognize programming languages too
![image](https://user-images.githubusercontent.com/15615524/56073623-e8691d00-5dd9-11e9-93d2-64b450d51d11.png)

**Comment by gandarez** [comment]
@alanhamlett any update here?

**Comment by gandarez** [comment]
Updated the issue with a comment https://github.com/pypa/setuptools/issues/1722#issuecomment-508067686

**Comment by codinggirl** [comment]
This issue need more information. Maybe you can update python lib setuptools, then rerun VS Code, and paste the wakatime debug log again, to see if anything useful.

**Comment by alanhamlett** [comment]
Might be fixed in v7.0.0 of the vscode extension, now that we're using the new [wakatime-cli](https://github.com/wakatime/wakatime-cli) that no longer needs Python.

**Comment by alanhamlett** [comment]
Closing since no activity lately.
